Quickstart 40

baranowb opened the pull request:
Initial quickstart 40 src.

sgilda commented:
Great example! 

The quickstart works with the sample XML file you provide, however, I'm getting NullPointerExceptions with most of the other XML files I have lying around and there is no feedback provided to the page. 

Is this supposed to work with any well-formatted XML file or is it specific to the one you provide?

sgilda commented:
Never mind. I see the file has to comply with the schema. :-)

baranowb commented:
Only specific at the moment, took me quite a lot of time to handle upload properly with JSF and Servlet only, so I did not want to complicate it even more.  In general it should not be hard to make it work for any file, with some small hacks.

sgilda commented:
Understood! You even say that in the page. 

One suggestion I have is to maybe highlight in the instructions and the README that it demonstrates both SAX and DOM parsing. Other than the drop-down, it's not really obvious that this example does this and it's a really cool feature.

On the web page, you provide a link to the example catalog.xml file. Is it also possible to provide the file location on disk in the instructions on the page? The JAR location is different from the file directory location, and it might save them some time finding the file to upload.

For example (where 'here' is your current link to the file in the JAR):
An example XML file can be found here and on the file system at this location: QUICKSTART_HOME/xml-jaxp/src/main/webapp/xml/catalog.xml. 
The example XSD file can be found here and on the file system at this location:  QUICKSTART_HOME/xml-jaxp/src/main/webapp/xsd/catalog.xsd.

Then, in the instructions, maybe say:
Please specify a file that complies to the catalog schema above, choose SAX or DOM for the parser, and click Upload.

Pete, I tested this for usability and it works great. If you are OK with the code, I think it's good to go.

baranowb commented:
Incorporated. If src is fine, will clone it to quickstart 41.

sgilda commented:
This looks great! You (or we) need to add 'Author: Bartosz Baranowski' under the title in the README file though.
Pete, have you reviewed the code?

pmuir commented on xml-jaxp/pom.xml:
```diff
@@ -0,0 +1,141 @@
+		</dependency>
+
+		<!-- dependency> <groupId>org.jboss.spec.javax.servlet</groupId> <artifactId>jboss-servlet-api_3.0_spec</artifactId> 
+			<scope>provided</scope> </dependency -->
```
Can we just remove this commented out code?

pmuir commented on xml-jaxp/pom.xml:
```diff
@@ -0,0 +1,141 @@
+        	<groupId>org.jboss.spec.javax.servlet</groupId>
+         	<artifactId>jboss-servlet-api_3.0_spec</artifactId>
+         	<scope>provided</scope>
+      	</dependency>
```
The whitespace is wrong here.

pmuir commented on xml-jaxp/pom.xml:
```diff
@@ -0,0 +1,141 @@
+			<artifactId>jboss-jaxrs-api_1.1_spec</artifactId>
+			<scope>provided</scope>
+		</dependency>
+		<!-- Import the JSF API, we use provided scope as the API is included in 
```
I think we need a new line here.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/catalog/Book.java:
```diff
@@ -0,0 +1,55 @@
+
+import java.util.Date;
+/**
+ * Simple java Bean which is filled by parser. It is used to abstract actual technology/library used to parse XML.
```
"JavaBean" not "java Bean". Probably better to just call it a Java object.

pmuir replied:
Also, I don't think "It is used to abstract actual technology/library used to parse XML." is very clear. It doesn't abstract anything, in and of itself...

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/catalog/Book.java:
```diff
@@ -0,0 +1,55 @@
+    }
+    public void setDescription(String description) {
+        this.description = description;
+    }
```
Can we get some new lines in here between methods and so on. The code formatter template should do this for you.

baranowb replied:
Hmm looking at all those formatting errors it seems I did not call formatter and upon commit, git changed formatting characters ( just a hunch, since in last quickstart file was fine, in git I got "format this" comments :) )

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadBean.java:
```diff
@@ -0,0 +1,137 @@
+ * 
+ */
+
+// Session
```
What is this comment (Session) for?

lincolnthree commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadBean.java:
```diff
@@ -0,0 +1,137 @@
+            this.parserFactory = (XMLParserFactory) this.getClass().forName(this.getParserFactoryClass()).newInstance();
+            this.catalog = this.parserFactory.instance().parse(is);
+        } catch (Exception e) {
+            // TODO Auto-generated catch block
```
Do you really want to swallow this exception? Shouldn't it be wrapped in RuntimeException (or the like) and re-thrown?

lincolnthree commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+            return;
+        }
+        
+        // create redirect URL.
```
Consider replacing these comments with a single multi-line / block comment?

lincolnthree commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+
+        }
+
+        // send 302 with default location, ie. 'https://localhost:8080/jboss-as-xml-jaxp', this will cause
```
same, block comment here instead?

lincolnthree commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+        int index = requestURL.lastIndexOf("/");
+        StringBuffer redirectLocation = req.getRequestURL().replace(index, requestURL.length(), "");
+
+        // fetch session scoped managed bean. Since this request wont be consumed by FacesServlet
```
block comment?

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadBean.java:
```diff
@@ -0,0 +1,137 @@
+
+    // name of form fields which are looked up in UploadServlet
+    private String inputName = "form:reg:file";
+    private String factoryName = "form:reg:factory";
```
Shouldn't this be on UploadServlet? Putting it on this controller class seems like bad separation of concerns to me.

pmuir replied:
Why not put these in statics? This is part of the app that is hardcoded.

lincolnthree commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/dom/DOMXMLParser.java:
```diff
@@ -0,0 +1,160 @@
+                    b.setPublishDate(d);
+                } catch (ParseException e) {
+                    // TODO Auto-generated catch block
+                    e.printStackTrace();
```
Again, maybe explain why we are swallowing this exception, or re-throw it?

lincolnthree commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/sax/SAXXMLParser.java:
```diff
@@ -0,0 +1,107 @@
+                    d = DATE_FORMATTER.parse(this.currentElementValue);
+                    this.book.setPublishDate(d);
+                } catch (ParseException e) {
+                    // TODO Auto-generated catch block
```
Explain or re-throw?

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Simple servlet to handle multipart file upload. By default, JSF 2.0 does not handle 'multipart/form-data' requests. <h:form../> does
+ * not even generate such requests.Frameworks like RichFaces come with Extension filters, which handle that. Since this is pure JSF 2.0, Servlet 3.0 and HTML
+ * example, we need to handle upload properly.
```
This is quite a "negative" explanation. Instead of explaining the limitations, stick to facts.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+ * not even generate such requests.Frameworks like RichFaces come with Extension filters, which handle that. Since this is pure JSF 2.0, Servlet 3.0 and HTML
+ * example, we need to handle upload properly.
+ * <br>
+ * This servlet register itself as handler for '/upload' context of application. When request is made with said context, this
```
Context is overused in java, I would describe this as the path.

baranowb replied:
ok

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+     * 
+     */
+    public FileUploadServlet() {
+        // TODO Auto-generated constructor stub
```
Can we remove the constructor if it's not in use.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+        //header in most case contains more than value, some parameters are present
+        //so make a check with String.contains(String).
+        if(!reqContentType.contains("multipart/form-data"))
+        {
```
We seem to have switch bracket styles here.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+        
+        // create redirect URL.
+        // Since @WebServlet(urlPatterns = { "/upload" }) makes this servlet recipent of request.
+        // FacesServlet wont see this request and once it is consumed browser will show blank page.
```
I think it might be simpler to just say that you need to redirect to a confirmation page or to the home page.

baranowb replied:
ok

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+        StringBuffer requestURL = req.getRequestURL();
+        //strip '/upload'
+        int index = requestURL.lastIndexOf("/");
+        StringBuffer redirectLocation = req.getRequestURL().replace(index, requestURL.length(), "");
```
A simpler pattern would be to ask the servlet context for the context root and then redirect to that.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+
+        // fetch session scoped managed bean. Since this request wont be consumed by FacesServlet
+        // this servlet must fetch bean and input some data.
+        FileUploadBean fileUpload = (FileUploadBean) req.getSession().getAttribute("fileUploadBean");
```
Why not just inject this?

baranowb replied:
IIRC I faced two beens beeing created at some point, I will double check.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+            resp.sendError(500,"Received request multipart request, but no file upload bean is available in session!");
+            //end processing.
+            return;
+        }
```
If you inject, you can get rid of this logic.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,147 @@
+                        && partContentType.equals("text/xml") 
+                        && partName != null
+                        //check if name matches to <input type="file".../> from FORM.
+                        && partName.equals(fileUpload.getInputName())) {
```
This needs to run through a formatter.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/dom/DOMXMLParser.java:
```diff
@@ -0,0 +1,160 @@
+
+        this.builder = factory.newDocumentBuilder();
+
+        // JAXP DOM(W3C) seems to use... SAX
```
This comment doesn't fill me with confidence. This quickstart is supposed to have been authored by an expert ;-)

pmuir replied:
Instead, say that you are using a SAX error handler...

baranowb replied:
Its just a disappointment comment :), will redo and inject something like 'wow, its cool' :)

pmuir commented:
Lot's of comments on the source. Generally, I'm very unhappy with the approach of setting up the parser via the web form etc. Why not use CDI alternatives for this, and instruct people how to make the switch at deployment. This will make the code simpler and more correct (specifying class names from the web is ripe for security issues).

pmuir commented:
Re Sande's comment on the schema. An NPE if we don't comply with the schema is not a good thing to happen. We should validate the uploaded file against the provide schema.

baranowb commented:
wrt to parser - Im fine with CDI, thought that ability to change on call will be better.

pmuir commented:
Regarding chaging the parser from the UI - I've never come across a real use case for doing something like that.

sgilda commented:
Hi Bartosz,
Where does this one stand? Is it ready to retest?
Thanks,
Sande

baranowb commented:
Starting to incorporate all comments, had some other tasks to do recently. I will post comment once Im done.

baranowb commented:
I will try to wrap up it tomorrow. Got tangled in git rebase hell... and I lost.

sgilda commented:
I know what you mean. Been there, done that ;-).

On 02/07/2012 05:02 PM, Bartosz Baranowski wrote:

> I will try to wrap up it tomorrow. Got tangled in git rebase hell... and I lost.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jbossas/quickstart/pull/74#issuecomment-3857741

sgilda commented:
I'm not a 'git' expert at all, but when I checkout your branch, it looks good to me:
   $ git fetch baranowb 
remote: Counting objects: 62, done.
remote: Compressing objects: 100% (11/11), done.
remote: Total 32 (delta 13), reused 32 (delta 13)
Unpacking objects: 100% (32/32), done.
From https://github.com/baranowb/quickstart
   0e23dde..5ebfb11  master     -> baranowb/master
   2d43a86..aae314c  quickstart-40 -> baranowb/quickstart-40

```
$ git checkout xml-jaxp 
```

Switched to branch 'xml-jaxp'
Your branch is behind 'baranowb/quickstart-40' by 1 commit, and can be fast-forwarded.

```
$ git reset --hard baranowb/quickstart-40
```

HEAD is now at aae314c incorporate propsed changes

Did you fetch, and do a reset?

sgilda commented:
Bartosz, did you ever figure this out?
Pete, do you understand what the problem with his  branch is?

baranowb commented:
Hi Sande
I am away till next week and have quite poor connection speed and quality so I wasnt able to work on tasks I did not have on my ddrive, I should be back by end of week. Will try to fix this on sunday so its ready for merge by monday. Sorry for delay.

sgilda commented:
No worries. I just didn't want you to think I was ignoring you! ;-)

On 02/17/2012 03:41 PM, Bartosz Baranowski wrote:

> Hi Sande
> I am away till next week and have quite poor connection speed and quality so I wasnt able to work on tasks I did not have on my ddrive, I should be back by end of week. Will try to fix this on sunday so its ready for merge by monday. Sorry for delay.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/jbossas/quickstart/pull/74#issuecomment-4028372

baranowb commented:
Not sure, when I checkout I have change in kitchensink-html5.../README.html file.. trying to fix that now.

baranowb commented:
baranowb@PORT1 /e/git/quickstart/kitchensink-html5-mobile (master)
$ git status
 On branch master
 Changes not staged for commit:
   (use "git add <file>..." to update what will be committed)
   (use "git checkout -- <file>..." to discard changes in working directory)

```
   modified:   readme.html
```

no changes added to commit (use "git add" and/or "git commit -a")

baranowb@PORT1 /e/git/quickstart/kitchensink-html5-mobile (master)
$ ls
LICENSE.txt  README.html  README.md  pom.xml  src

baranowb commented:
Changes that block my master are .... well, those were fetched when I ran fetch/pull commands against my old master to bring it up to date.

Just compared current quickstart master. Makes no sense, 'git diff' shows:
 -&lt;h1>html5-mobile: Mobile Example Using Plain Old HTML5 (POH5)</h1>
but this line is present in current master.

baranowb commented:
Seems like file rename screwed my pull request:

$ git status
 On branch master
 Your branch is ahead of 'origin/master' by 3 commits.

 Changes not staged for commit:
   (use "git add <file>..." to update what will be committed)
   (use "git checkout -- <file>..." to discard changes in working directory)

```
   modified:   kitchensink-html5-mobile/readme.html
```

no changes added to commit (use "git add" and/or "git commit -a")

baranowb@PORT1 /e/git/quickstart (master)
$ cd kitchensink-html5-mobile/

baranowb@PORT1 /e/git/quickstart/kitchensink-html5-mobile (master)
$ ls
LICENSE.txt  README.html  README.md  pom.xml  src

baranowb@PORT1 /e/git/quickstart/kitchensink-html5-mobile (master)
$ git rm -f readme.html
rm 'kitchensink-html5-mobile/README.html'

baranowb@PORT1 /e/git/quickstart/kitchensink-html5-mobile (master)
$ git status
 On branch master
 Your branch is ahead of 'origin/master' by 3 commits.

 Changes to be committed:
   (use "git reset HEAD <file>..." to unstage)

```
   deleted:    README.html
```

 Changes not staged for commit:
   (use "git add/rm <file>..." to update what will be committed)
   (use "git checkout -- <file>..." to discard changes in working directory)

```
   deleted:    readme.html
```

baranowb commented:
Ok, finally managed to fix my master, rephrased two comments.

sgilda commented:
Thanks Bartosz. I'll try to get to it today.

sgilda commented:
Hi Bartosz,
This really looks great! One suggestion I have is that when someone uploads an XML file that doesn't match the spec, instead of displaying the "HTTP Status 500 - " and the stacktrace, it might be nicer to catch the exception and redisplay the original page with an error message at the top that says something like 
"The file you uploaded. <if you can put the filename here, that would be nice>, does not match the schema specified in QUICKSTART_HOME/xml-jaxp/src/main/webapp/xsd/catalog.xsd. If you do not want to create a file that matches the schema, you can test this quickstart by uploading the file QUICKSTART_HOME/xml-jaxp/src/main/webapp/xml/catalog.xml".

sgilda commented:
This looks much better. I think this will be more useful for developers since you usually don't want to throw exceptions to the page. The only suggestions I have at this point:
- When it fails, clear out the previous XML file content since it's no longer valid. For example, I uploaded the correct file, then I tried another XML file that didn't match the schema. It redisplayed the content from the previous request, followed by "Schema validation failed".
- Could the error message "<h:messages errorClass="invalid"/>" be displayed in red so it's obvious it's an error?

Pete, I'm not qualified to review the "hack". Can you do that? Or is there someone I could ask to do that review?

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadBean.java:
```diff
@@ -0,0 +1,137 @@
+    // parserFactory. Annotated as 'ManagedProperty' to make container inject value. 'value' is EL expresion which will make
+    // container inject managed bean. Actual implementation of bean is beeing set by faces-config.xml entry.
+    @ManagedProperty(value = "#{parserFactory}")
+    XMLParserFactory parserFactory;
```
I thought we agreed to use CDI alternatives for this? Anyway, we shouldn't be using JSF managed beans in this example.

baranowb replied:
Hmm missed that. Not very faimiliar with CDI though. I will change it than.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadBean.java:
```diff
@@ -0,0 +1,137 @@
+     * @return
+     */
+    public XMLParserFactory getParserFactory() {
+        return parserFactory;
```
Rather than use a factory, we should use a producer method.

baranowb replied:
Ok

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadBean.java:
```diff
@@ -0,0 +1,137 @@
+ * 
+ */
+@SessionScoped
+@ManagedBean
```
We shouldn't be using @ManagedBean at all in a CDI example.

pmuir commented on xml-jaxp/pom.xml:
```diff
@@ -0,0 +1,138 @@
+	<version>7.0.2-SNAPSHOT</version>
+	<packaging>war</packaging>
+	<name>JBoss AS Quickstarts: Hiberanate3</name>
+	<description>JBoss AS Quickstarts: Hiberanate3</description>
```
This isn't the Hibernate 3 example ;-)

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/ErrorPhaseListener.java:
```diff
@@ -0,0 +1,91 @@
+ * @author baranowb
+ * 
+ */
+public class ErrorPhaseListener implements PhaseListener {
```
What is this about? No javadoc.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,148 @@
+
+        // fetch session scoped managed bean. Since this request wont be consumed by FacesServlet
+        // this servlet must fetch bean and input some data.
+        final FileUploadBean fileUpload = (FileUploadBean) req.getSession().getAttribute("fileUploadBean");
```
We should use CDI to inject a CDI bean, not this hacky lookup from the session which relies on JSF implementation details.

baranowb replied:
Possibly I did something wrong here - if I use CDI here I get two different instances of FileUploadBean - one created when it is injected into servlet, second created by JSF.

pmuir replied:
Did you make the FileUploadBean request scoped?

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,148 @@
+        //even though we can create FacesContext its does not transit to JSF
+        final ServletContext servletContext = super.getServletContext();
+        if(servletContext.getAttribute(MESSAGES_ATTRIBUTE)==null)
+        {
```
Wrong bracket style here.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,148 @@
+        final ServletContext servletContext = super.getServletContext();
+        if(servletContext.getAttribute(MESSAGES_ATTRIBUTE)==null)
+        {
+            servletContext.setAttribute(MESSAGES_ATTRIBUTE, new ArrayList<FacesMessage>());
```
Again, better to use CDI here, rather than fiddling around with the session.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/FileUploadServlet.java:
```diff
@@ -0,0 +1,148 @@
+                    LifecycleFactory lifecycleFactory = (LifecycleFactory) FactoryFinder.getFactory(FactoryFinder.LIFECYCLE_FACTORY);
+                    Lifecycle lifecycle = lifecycleFactory.getLifecycle(LifecycleFactory.DEFAULT_LIFECYCLE);
+                    FacesContext facesContext = contextFactory.getFacesContext(servletContext, req, resp, lifecycle);
+                    fileUpload.parseUpload(is,schemaURL);
```
This is definitely wrong. What are you doing here?

baranowb replied:
FacesContext is not accessible in Servlet, hence its not possible to queue FacesMessages as Sande asked. Googled a bit and this was one way to solve this, atleast partially.

pmuir replied:
Ok, but this is really a horrible thing to do.

It would be better to create some FacesMessages here, store them in the CDI request scope, and then enqueue them once JSF is available. It might be even better to just grap the input stream at this point, and stick it in the request scope. You could then parse it inside the JSF action method.

baranowb replied:
Roger. I didnt like this hack, told that Sande just before commit :)

pmuir commented on xml-jaxp/src/main/webapp/WEB-INF/faces-config.xml:
```diff
@@ -0,0 +1,28 @@
+		<managed-bean-class>org.jboss.as.quickstart.xml.jaxp.sax.SAXXMLParserFactory</managed-bean-class>
+		<managed-bean-scope>session</managed-bean-scope>
+	</managed-bean>
+	 -->
```
As discussed, this should be CDI alternatives.

pmuir commented:
Some more comments added.

We should be using CDI, not JSF managed beans here. There are also some very odd hacks around moving error messages around. Why do we need these?

baranowb committed:
new brancy for q40, since old one was screwed by me.

sgilda commented:
Hi Bartosz,

I just tested this out and it builds, deploys, and runs file. 
I do have a few minor  comments:
--  I don't see any errors displayed on the page when I try to load an XML file (for example, the pom.xml file) that doesn't match the schema. I can see that you have error handling code to accumulate errors and display errorHolder.errorMessages on the page, but it doesn't appear to be working.
--  A couple of the pom.xml versions need to be updated to match the AS 7 versions here: https://docs.jboss.org/author/display/AS71/Maven+POM+Versions+Checklist
--  See Lincoln's comment about replacing // comments with a single multi-line block comment. I still see a lot of code comments that haven't been blocked /\* ... */ and it's much easier to read.

Pete or Lincoln, do you want to review the latest code?

Thanks, Sande

baranowb commented:
Weird, I did test just before commit and it worked for me ;[ Wrt to Lincoln comment - thought it was just to have comments gathered up together so they are no scattered per line.

baranowb committed:
add block comments, fix how error is displayed

sgilda commented:
I tested this and it works great now. If I try to upload a file that doesn't match the schema, it gives me an informative error 
Message: Failed to validate file against schema!    Exception: cvc-elt.1: Cannot find the declaration of element 'project'.

Pete or Lincoln, do you want to review this code one more time to make sure your concerns are addressed?

Thanks, Sande

baranowb commented:
It was my bad, removed one line to many when I was cleaning code before commit.

sgilda commented:
Pete, this one works for me. Do you need to do a code review?

pmuir commented:
I'll do one, hopefully today.

pmuir commented on xml-jaxp/src/main/java/org/jboss/as/quickstart/xml/jaxp/dom/DOMXMLParser.java:
```diff
@@ -0,0 +1,172 @@
+ */
+@SuppressWarnings("unchecked")
+@RequestScoped
+@Default
```
This @Default is uneeded, I'll remove it before merge.

pmuir commented on xml-jaxp/src/main/webapp/WEB-INF/jboss-deployment-structure.xml:
```diff
@@ -0,0 +1,9 @@
+	<deployment>
+		<dependencies>
+			
+			<module name="org.apache.commons.io" />
```
This should have some comments, could you send a pull after merge for this?

pmuir commented:
This is replaced by https://github.com/jbossas/quickstart/pull/152 - please review there.
